network/p2pNetwork.go

```diff
@@ -0,0 +1,453 @@
+		config:                 cfg,
+		broadcastQueueHighPrio: make(chan broadcastRequest, outgoingMessagesBufferSize),
+		broadcastQueueBulk:     make(chan broadcastRequest, 100),
+	}
+
+	net.service, err = p2p.MakeService(net.ctx, log, cfg, datadir, pstore, net.wsStreamHandler)
+	if err != nil {
+		return nil, err
+	}
+
+	err = net.setup()
+	if err != nil {
+		return nil, err
+	}
+
+	return net, nil
+}
+
+func (n *P2PNetwork) setup() error {
+	if n.broadcaster.slowWritingPeerMonitorInterval == 0 {
+		n.broadcaster.slowWritingPeerMonitorInterval = slowWritingPeerMonitorInterval
+	}
+	return nil
+}
+
+// Start threads, listen on sockets.
+func (n *P2PNetwork) Start() {
+	n.wg.Add(1)
+	go n.txTopicHandleLoop()
+
+	if n.wsPeersConnectivityCheckTicker != nil {
+		n.wsPeersConnectivityCheckTicker.Stop()
+	}
+	n.wsPeersConnectivityCheckTicker = time.NewTicker(connectionActivityMonitorInterval)
+	for i := 0; i < incomingThreads; i++ {
+		n.wg.Add(1)
+		// We pass the peersConnectivityCheckTicker.C here so that we don't need to syncronize the access to the ticker's data structure.
+		go n.handler.messageHandlerThread(&n.wg, n.wsPeersConnectivityCheckTicker.C, n)
+	}
+
+	n.wg.Add(1)
+	go n.broadcaster.broadcastThread(&n.wg, n)
+	n.service.DialPeersUntilTargetCount(n.config.GossipFanout)
+
+	n.wg.Add(1)
+	go n.meshThread()
+}
+
+// Stop closes sockets and stop threads.
+func (n *P2PNetwork) Stop() {
+	n.handler.ClearHandlers([]Tag{})
+	if n.wsPeersConnectivityCheckTicker != nil {
+		n.wsPeersConnectivityCheckTicker.Stop()
+		n.wsPeersConnectivityCheckTicker = nil
+	}
+	n.ctxCancel()
+	n.service.Close()
+	n.wg.Wait()
+}
+
+func (n *P2PNetwork) meshThread() {
+	defer n.wg.Done()
+	timer := time.NewTicker(meshThreadInterval)
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			n.service.DialPeersUntilTargetCount(n.config.GossipFanout)
+		case <-n.ctx.Done():
+			return
+		}
+	}
+}
+
+// GetGenesisID implements GossipNode
+func (n *P2PNetwork) GetGenesisID() string {
+	return n.genesisID
+}
+
+// Address returns a string and whether that is a 'final' address or guessed.
+func (n *P2PNetwork) Address() (string, bool) {
+	addrInfo := n.service.AddrInfo()
+	addrs, err := peer.AddrInfoToP2pAddrs(&addrInfo)
+	if err != nil {
+		n.log.Warnf("Failed to generate valid multiaddr: %v", err)
+		return "", false
+	}
+	if len(addrs) == 0 {
+		return "", false
+	}
+	if len(addrs) > 1 {
+		n.log.Infof("Multiple addresses found, using first one from %v", addrs)
+	}
+
+	return addrs[0].String(), true
+}
+
+// Broadcast sends a message.
+func (n *P2PNetwork) Broadcast(ctx context.Context, tag protocol.Tag, data []byte, wait bool, except Peer) error {
+	// For tags using pubsub topics, publish to GossipSub
+	if topic, ok := n.topicTags[tag]; ok {
+		return n.service.Publish(ctx, topic, data)
+	}
+	// Otherwise broadcast over websocket protocol stream
+	return n.broadcaster.BroadcastArray(ctx, []protocol.Tag{tag}, [][]byte{data}, wait, except)
```

Comment:
this mean broadcast requests to p2pnet both published to gossib sub and pushed to neighbors?